model_free.py

```diff
@@ -0,0 +1,235 @@
+
+            if pendulum:
+                log['theta'].append(env.x[0])
+                log['thetadot'].append(env.x[1])
+
+            if done: break
+
+    # Subplots
+    fig, axs = plt.subplots(2, 1)
+    
+    if pendulum: fig, axs = plt.subplots(3, 1)
+
+    # Plot data and save to png file
+    axs[0].plot(np.arange(num_episodes), ep_reward)
+    axs[0].set_xlabel('Episodes')
+    axs[0].set_ylabel('Reward')
+    axs[0].grid()
+
+    axs[1].plot(log['t'], log['s'])
+    axs[1].plot(log['t'][:-1], log['a'])
+    axs[1].plot(log['t'][:-1], log['r'])
+    axs[1].set_xlabel('t')
+    axs[1].set_ylabel('State | Action | Reward')
+    axs[1].legend(['s', 'a', 'r'], loc="upper right")
+    axs[1].grid()
+
+    if pendulum: 
+        axs[2].plot(log['t'], log['theta'])
+        axs[2].plot(log['t'], log['thetadot'])
+        axs[2].set_xlabel('t')
+        axs[2].set_ylabel('Theta | ThetaDot')
+        axs[2].legend(['Theta', 'ThetaDot'])
+        axs[2].grid()
+    
+    fig.tight_layout()
+    
+    plt.savefig(path)
+
+    return Q
+
+def Q_learning(env, num_episodes, gamma, alpha, epsilon, path, pendulum):
+    # initialize Q matrix
+    Q = np.zeros((env.num_states, env.num_actions))
+
+    # rewards for plotting
+    ep_reward = np.zeros(num_episodes)
+
+    # loop through episodes
+    for episode in range(num_episodes):
+        # initialize s
+        state = env.reset()
+
+        # For plotting
+        # Create log to store data from simulation
+        log = {
+            't': [0],
+            's': [state],
+            'a': [],
+            'r': [],
+        }
+
+        if pendulum:
+            log['theta'] = [env.x[0]] 
+            log['thetadot'] = [env.x[1]]
+
+        # step through the environment
+        for t in itertools.count():
+            # take a step
+            action = epsilon_greedy_policy(Q, epsilon, state)
+            next_state, reward, done = env.step(action)
+
+            # choose next best action
+            best_next_action = np.argmax(Q[next_state])
+
+            # calculate Q value
+            Q[state, action] = Q[state, action] + alpha * (reward + gamma * Q[next_state, best_next_action] - Q[state, action])
+
+            # update state and action
+            state = next_state
+
+            # update values for plotting
+            ep_reward[episode] += reward
+
+            log['t'].append(log['t'][-1] + 1)
+            log['s'].append(state)
+            log['a'].append(action)
+            log['r'].append(reward)
+
+            if pendulum:
+                log['theta'].append(env.x[0])
+                log['thetadot'].append(env.x[1])
+
+            if done: break
+
+    # Subplots
+    fig, axs = plt.subplots(2, 1)
+
+    if pendulum: fig, axs = plt.subplots(3, 1)
+
+    # Plot data and save to png file
+    axs[0].plot(np.arange(num_episodes), ep_reward)
+    axs[0].set_xlabel('Episodes')
+    axs[0].set_ylabel('Reward')
+    axs[0].grid()
+
+    axs[1].plot(log['t'], log['s'])
```

Comment:
I think we might need to simulate and plot the trained agent following a purely greedy policy instead of the epsilon greedy we used in training. I'll ask on campuswire